src/include/common/perf_monitor.h

```diff
@@ -0,0 +1,203 @@
+     * Total cycles. Be wary of what happens during CPU frequency scaling.
+     */
+    uint64_t cpu_cycles_;
+    /**
+     * Retired instructions. Be careful, these can be affected by various issues, most notably hardware interrupt
+     * counts.
+     */
+    uint64_t instructions_;
+    /**
+     * Cache accesses. Usually this indicates Last Level Cache accesses but this may vary depending on your CPU.  This
+     * may include prefetches and coherency messages; again this depends on the design of your CPU.
+     */
+    uint64_t cache_references_;
+    /**
+     * Cache misses.  Usually this indicates Last Level Cache misses.
+     */
+    uint64_t cache_misses_;
+    // TODO(Matt): there seems to be a bug with enabling these counters along with the cache counters. When enabled,
+    // just get 0s out of all of the counters. Eventually we might want them but can't enable them right now.
+    // https://lkml.org/lkml/2018/2/13/810
+    // uint64_t branch_instructions_;
+    // uint64_t branch_misses_;
+    /**
+     * Bus cycles, which can be different from total cycles.
+     */
+    uint64_t bus_cycles_;
+    /**
+     * Total cycles; not affected by CPU frequency scaling.
+     */
+    uint64_t ref_cpu_cycles_;
+
+    /**
+     * compound assignment
+     * @param rhs you know subtraction? this is the right side of that binary operator
+     * @return reference to this
+     */
+    PerfCounters &operator-=(const PerfCounters &rhs) {
+      TERRIER_ASSERT(this->num_events_ == rhs.num_events_,
+                     "Operating on PerfCounters objects with different num_events_ doesnt' make sense.");
+      this->cpu_cycles_ -= rhs.cpu_cycles_;
+      this->instructions_ -= rhs.instructions_;
+      this->cache_references_ -= rhs.cache_references_;
+      this->cache_misses_ -= rhs.cache_misses_;
+      // this->branch_instructions_ -= rhs.branch_instructions_;
+      // this->branch_misses_ -= rhs.branch_misses_;
+      this->bus_cycles_ -= rhs.bus_cycles_;
+      this->ref_cpu_cycles_ -= rhs.ref_cpu_cycles_;
+      return *this;
+    }
+
+    /**
+     * subtract implemented from compound assignment. passing lhs by value helps optimize chained a+b+c
+     * @param lhs you know subtraction? this is the left side of that binary operator
+     * @param rhs you know subtraction? this is the right side of that binary operator
+     * @return
+     */
+    friend PerfCounters operator-(PerfCounters lhs, const PerfCounters &rhs) {
+      lhs -= rhs;
+      return lhs;
+    }
+  };
+
+  PerfMonitor() {
+#if __APPLE__
+    // Apple doesn't support perf events and currently doesn't expose an equivalent kernal API
+    valid_ = false;
+#else
+    // Initialize perf configuration
+    perf_event_attr pe;
+    std::memset(&pe, 0, sizeof(perf_event_attr));
+    pe.type = PERF_TYPE_HARDWARE;
+    pe.size = sizeof(perf_event_attr);
+    pe.disabled = 1;
+    pe.exclude_kernel = 1;
+    pe.exclude_hv = 1;
+    pe.read_format = PERF_FORMAT_GROUP;
```

Comment:
Can you explain what are these?

Reply:
I would generally just refer people to the man page that I link at the top of the file. I'm not sure how useful it is to reproduce a ton of the man page. I did it where absolutely necessary.